flowauth/frontend/src/util/util.js

```diff
@@ -26,80 +27,144 @@ function zip() {
   });
 }
 
-export function scopesGraph(array) {
-  const nested = {};
-  Object.keys(array).forEach((scope) => {
-    let obj = nested;
-    let last_split = [];
-    let parents = [];
-    scope.split("&").forEach((sub_scope) => {
-      obj["parent"] = parents.join("&");
-      let split = sub_scope.split(".");
-      split = zip(last_split, split)
-        .dropWhile((x) => x[0] !== x[1])
-        .map((x) => x[1])
-        .filter((x) => x !== undefined);
-      split.forEach((k, ix) => {
-        if (!(k in obj)) {
-          obj[k] = {};
-        }
-        obj = obj[k];
-        obj["parent"] = parents.join("&");
-      });
-      obj["full_path"] = sub_scope;
-      last_split = split;
-      parents.push(sub_scope);
-    });
-  });
-  return nested;
+
+
+
+
+  /** Takes an array of scopes, most of which are triplets in the form admin_level:top_level_query:dependent_query
+  * and converts them into a graph keyed on admin_level -> top_level_query -> dependent_query
+  */
+export function scopesGraph(scopes_obj) {
+  var scopes_array
+  if (!Array.isArray(scopes_obj)){
+    scopes_array = Object.keys(scopes_obj)
+  }
+  else{
+    scopes_array = scopes_obj
+  }
+  const triplet_scopes = scopes_array.filter(row => row.includes(":"))  // It's this or a regex
+  const split_scopes = triplet_scopes.map(row => row.split(":"))
+  const scopes_tree = {}
+  for (const row of split_scopes){
+    var [ admin_level, tl_query, inner_query ] = row
+
+    if (!(Object.keys(scopes_tree).includes(admin_level))){
+      scopes_tree[admin_level] = {}
+    }
+    if (!(Object.keys(scopes_tree[admin_level]).includes(tl_query))){
+      scopes_tree[admin_level][tl_query] = {}
+    }
+    scopes_tree[admin_level][tl_query][inner_query] = true
```

Comment:
Can this be implemented to handle ':'-delimited scopes with an arbitrary number of parts? If, in the future, we decided to add some 2-part scopes (or 4-part, etc.), it would be nice if FlowAuth was already able to handle these - then FlowAPI could include the new scopes in the API spec, and that API spec could be uploaded to an existing FlowAuth, without requiring a FlowAuth re-deployment.